feat: per-enrollment data

gkc opened the pull request:
**- What I did**
- feat: allow enrollments read and write access to the `<enrollmentId>.a.__e@alice` namespace
- feat: ensure that data in that namespace is moved to `.r.__e@` or `.d.__e@` when enrollments are revoked or deleted
- feat: ensure that data in that namespace is moved back to `.a.__e@` if a revoked enrollment is unrevoked
- feat: to enable the above, added facility to add `preRemove` and `postRemove` to a `WritableKeyStore`
- fix: ensure there is no orphaned data when enrollments are deleted (or expire and are deleted)
- test: added many tests for all of the above
- test: also fixed some other tests
- fix: fixed sync for sub-namespaces
- refactor: a shedload of refactoring (to reviewers: my apologies for the amount of this)
- feat: added caching to EnrollmentManager since we are fetching enrollment details on every request, and enrollments are now only accessed via the EnrollmentManager

**- How I did it**
See change-set and comments thereon. All material changes are called out in a comment. Everything else was refactoring.

**- How to verify it**
Tests pass

gkc committed:
feat: allow enrollments rw access to the `<enrollmentId>.a.__e` namespace

gkc committed:
chore: add `strict_top_level_inference: false` to analysis_options.yaml as fixing all of it is a cascade of busywork some of which could be dangerous. Have made a ticket to deal with this ... later.

gkc committed:
refactor: removed some dead code which I noticed while doing something else and which looked like a bug but turned out was just obsolete code that hadn't been removed when it should have

gkc committed:
refactor: added EnrollmentConstants to hold the constants in enroll_constants.dart
refactor: minimize gets of the singleton in EnrollVerbHandler

gkc committed:
refactor: moved a function from AbstractVerbHandler to EnrollmentManager
fix: restored `newEnrollmentKeyPattern` to EnrollmentConstants as it was being used not just (incorrectly) as a regex but also in key construction
refactor: moved `_fetchAllEnrollments` function from EnrollVerbHandler to EnrollmentManager and renamed to `fetchEnrollments`

gkc committed:
Merge remote-tracking branch 'origin/trunk' into trunk

gkc committed:
feat: added `getEnrollDataStoreValueById` function to EnrollmentManager
refactor: made `enrollmentKeysList` an optional parameter to `EnrollmentManager.fetchEnrollments`

gkc committed:
refactor: moved the deletion of enrollments' encryption keys into the EnrollmentManager.remove method
refactor: make EnrollmentManager constructor require atSign rather than using `AtSecondaryServerImpl.getInstance().currentAtSign`
chore: standardize various things in tests

gkc committed:
refactor: renamed EnrollmentManager.get to getEnrollment
fix: while looking at everywhere that EnrollmentManager.getEnrollment is used, found a bug in the SyncProgressVerbHandler which wasn't using the standard `isAuthorized` check and would therefore miss out on situations where for example an enrollment has rw access to the "foo.bar" but not to the "bar" namespace, it would not receive data like x.bar and y.bar
TODO: Need to add unit tests for the above sync fix
TODO: Need to look at caching in EnrollmentManager.getEnrollment so we're not constantly reading from datastore

gkc committed:
chore: dart format, and reverted some logging changes

gkc committed:
chore: cleaned up noise in unit test output

gkc committed:
chore: reuse test_utils in sync_unit_test.dart

gkc committed:
test: added test for sync of sub-namespaces
refactor: (test) lots of test refactoring

gkc committed:
test: added some assertions to the enrollment unit tests
refactor: further refactoring of EnrollmentManager

gkc committed:
refactor: EnrollVerbHandler now uses EnrollDataStoreValue.toJsonExtended instead of itself creating the json map

gkc committed:
test: Fixed the unit test which claimed to deal with expired enrollments but in fact did not
feat: Upon startup the server will now remove all legacy Apkam public key records from the keyStore.
refactor: EnrollmentManager polishing

gkc committed:
feat: throw appropriate exceptions in various enrollment-related code, instead of AtEnrollmentException, which is not in the exception handling map in at_commons and therefore gets treated as an InternalServerException by the atServer's GlobalExceptionHandler

gkc committed:
test: remove obsolete functional test, and change expectations of  "Internal server exception" to "Illegal state exception" as required

gkc committed:
test: (functional tests) change expectations of "Illegal state exception" to "UnAuthorized client in request" as required

gkc committed:
fix: oops

gkc committed:
fix: remove orphaned APKAM encryption keys on startup
fix: ensure there are no more orphaned APKAM encryption keys from enrollment expirations
refactor: lots of refactoring to remove duplication and stuff

gkc committed:
fix: change log level to INFO in BaseResponseHandler when logging error responses

gkc committed:
test: tests to cover datastore consistency for enrollment and enrollment-related records.

gkc committed:
Merge remote-tracking branch 'origin/trunk' into per-enrollment-namespaces

gkc committed:
test: adjust wait in a unit test to prevent flaking

gkc committed:
feat: preRemove and postRemove hooks for WritableKeyStore. Allows EnrollmentManager to register an interest in keys being removed and then ensure that if an enrollment is being removed (whether deliberately or because it has expired) we can consistently handle all the other things that need to happen
feat: per-enrollment namespace lookups handling expired enrollments which have not yet been harvested by the key expiry job

gkc committed:
Merge remote-tracking branch 'origin/trunk' into per-enrollment-namespaces

gkc committed:
test: implemented last set of unit tests for per-enrollment namespace data (dynamic checking of enrollment status when handling lookups of things in .a.__e@

gkc commented on packages/at_persistence_secondary_server/CHANGELOG.md:
```diff
@@ -1,3 +1,6 @@
+## 4.1.0
+- feat: Add `preRemoveHook` and `postRemoveHook` to KeyStore interfaces
```
Some small additions to the at_persistence_spec and at_persistence_secondary_server for this. And why? Well, at_secondary_server code needed some control over what happens when deleting expired keys - specifically when deleting expired enrollments, so that per-enrollment data can be moved from the `.a.__e@` to the `.d.__e@` namespace. However, (1) `deleteExpiredKeys` is in the at_persistence_secondary_server package and (2) code to handle active deletion of enrollments could be easily de-dup'd into the same `preRemove` hook

gkc commented on packages/at_secondary_server/lib/src/constants/enroll_constants.dart:
```diff
@@ -1,5 +1,22 @@
+  static const String perEnrollmentDeleted = 'd.__e';
+  static const String regexForPerEnrollmentNamespaces =
+      '\\.(?<EnId>[^\\.]+)\\.[ard]\\.__e@';
+}
```
I will move this to at_commons before finalizing this PR

gkc commented on packages/at_secondary_server/lib/src/enroll/enroll_datastore_value.dart:
```diff
@@ -25,6 +25,13 @@ class EnrollDataStoreValue {
+    m['status'] = approval?.state;
+    m['namespace'] = m['namespaces'];
+    return m;
+  }
```
the handling of enrollments JSON was a bit wack ... this helped clean it up a little

gkc commented on packages/at_secondary_server/lib/src/enroll/enrollment_manager.dart:
```diff
@@ -4,26 +4,28 @@ import 'package:at_commons/at_commons.dart';
   ///
   /// The [keyStore] is required to interact with the persistence layer.
-  EnrollmentManager(this._keyStore);
+  EnrollmentManager(this.keyStore, this.atSign);
```
The primary changes to implement the per-enrollment-data feature are here in EnrollmentManager.

gkc commented on packages/at_secondary_server/lib/src/server/at_secondary_impl.dart:
```diff
@@ -172,7 +173,29 @@ class AtSecondaryServerImpl implements AtSecondaryServer {
+    if (deletedKeys.isNotEmpty) {
+      logger.info('Removed orphaned APKAM encryption keys: $deletedKeys');
+    }
+
```
The code below here was moved to here for improved readability

gkc commented on packages/at_secondary_server/lib/src/server/at_secondary_impl.dart:
```diff
@@ -172,7 +173,29 @@ class AtSecondaryServerImpl implements AtSecondaryServer {
+    deletedKeys = await enrollmentManager.removeOrphanedApkamEncryptionKeys();
+    if (deletedKeys.isNotEmpty) {
+      logger.info('Removed orphaned APKAM encryption keys: $deletedKeys');
+    }
```
The code above (1) removes legacy APKAM public keys which we do not need and (2) removes APKAM encryption keys which were orphaned after APKAM expiries (part of this PR ensures that there are no more orphans in the future)

gkc commented on packages/at_secondary_server/lib/src/verb/handler/abstract_verb_handler.dart:
```diff
@@ -140,23 +142,8 @@ abstract class AbstractVerbHandler implements VerbHandler {
-    } on KeyNotFoundException {
-      logger.severe('$enrollmentKey does not exist in the keystore');
-      rethrow;
-    }
```
Moved the above function to EnrollmentManager

gkc commented on packages/at_secondary_server/lib/src/verb/handler/enroll_verb_handler.dart:
```diff
@@ -216,21 +244,9 @@ class EnrollVerbHandler extends AbstractVerbHandler {
-      delayForInvalidOTPSeries.clear();
-      delayForInvalidOTPSeries.addAll([0, initialDelayInMilliseconds]);
-    }
-
```
The changes above (addition of code, removal of code) re delayForInvalidOTPSeries were to correctly implement the desired behaviour)

gkc commented on packages/at_secondary_server/lib/src/verb/handler/enroll_verb_handler.dart:
```diff
@@ -31,15 +32,15 @@ class EnrollVerbHandler extends AbstractVerbHandler {
-  EnrollVerbHandler(super.keyStore);
+  final EnrollmentManager enMgr;
+
+  EnrollVerbHandler(super.keyStore, this.enMgr);
```
Most of the changes here are refactoring. Where they are significant, I will add comment.

gkc commented on packages/at_secondary_server/lib/src/verb/handler/enroll_verb_handler.dart:
```diff
@@ -618,79 +610,57 @@ class EnrollVerbHandler extends AbstractVerbHandler {
-    await AtSecondaryServerImpl.getInstance()
-        .enrollmentManager
-        .put(enrollmentId, atData);
-  }
```
Inlined the function for readability as only used in one place

gkc commented on packages/at_secondary_server/lib/src/verb/handler/local_lookup_verb_handler.dart:
```diff
@@ -76,6 +77,22 @@ class LocalLookupVerbHandler extends AbstractVerbHandler {
+      } on KeyNotFoundException {
+        // We don't need to do anything more
+      }
+    }
```
This takes care of the tricky edge case where an enrollment has expired but has not yet been cleaned up by the `deleteExpiredKeys` job. By trying to fetch the enrollment, EnrollmentManager will detect that it has expired and will do the appropriate cleanup etc.

gkc commented on packages/at_secondary_server/lib/src/verb/handler/lookup_verb_handler.dart:
```diff
@@ -53,6 +58,21 @@ class LookupVerbHandler extends AbstractVerbHandler {
+      } on KeyNotFoundException {
+        // We don't need to do anything more
+      }
+    }
```
Same change as in LocalLookupVerbHandler above

gkc committed:
docs: corrected code comment in ProxyLookupVerbHandler

gkc committed:
chore: removed commented-out old code

gkc commented on packages/at_secondary_server/lib/src/verb/handler/sync_progressive_verb_handler.dart:
```diff
@@ -93,20 +91,17 @@ class SyncProgressiveVerbHandler extends AbstractVerbHandler {
+      if (!(await isAuthorized(
+        connectionMetadata,
+        atKey: commitEntryIterator.current.key!,
+      ))) {
```
The changes above are to ensure that keys in sub-namespaces will sync correctly. (New tests have been added for this)

gkc commented on packages/at_secondary_server/test/enroll_verb_test.dart:
```diff
@@ -3,12 +3,14 @@ import 'dart:convert';
 
 import 'package:at_commons/at_commons.dart';
```
All of the new unit tests for the per-enrollment-data feature, and for data consistency checking, are here

gkc commented on packages/at_secondary_server/test/pkam_verb_test.dart:
```diff
@@ -125,24 +129,47 @@ void main() {
           'enrollment_id: enrollId is denied');
     });
 
     test('verify apkam behaviour - case: enrollment expired ', () async {
```
This test wasn't actually testing enrollment expired. Now it does.

gkc commented on packages/at_secondary_server/test/sync_unit_test.dart:
```diff
@@ -975,124 +922,157 @@ void main() {
+        }
+        await handler.processVerb(response, syncVerbParams, inboundConnection);
+        return jsonDecode(response.data!);
+      }
```
New test: sync with sub-namespace permissions. Extended and renamed the existing test (namespace permissions)

gkc commented on tests/at_functional_test/test/enroll_verb_test.dart:
```diff
@@ -499,42 +499,6 @@ void main() {
           authType: AuthType.pkam);
     });
 
-    test('A test to verify pkam public key is stored in __pkams namespace',
```
Removed, because we don't do this any more

xavierchanth commented on packages/at_persistence_secondary_server/pubspec.yaml:
```diff
@@ -19,6 +19,10 @@ dependencies:
   at_persistence_spec: ^3.0.0
   meta: ^1.8.0
 
+dependency_overrides:
+  at_persistence_spec:
+    path: ../at_persistence_spec
+
```
Was this for testing or meant to be left in?

gkc replied:
@XavierChanth see notes at the beginning of the PR description

gkc replied:
@XavierChanth see notes at the beginning of the PR description

gkc committed:
refactor: Move EnrollmentConstants to at_commons

gkc committed:
Merge remote-tracking branch 'origin/trunk' into per-enrollment-namespaces

# Conflicts:
#	packages/at_persistence_spec/lib/src/keystore/keystore.dart

gkc committed:
build: removed dependency overrides

gkc committed:
feat: Added a cache to EnrollmentManager since enrollments are fetched so often - at least once, frequently twice, and sometimes many times (scan, sync, ...) for each verb request

gkc commented:
@XavierChanth thanks for review - since then I have
- moved EnrollmentConstants to at_commons ([commit](https://github.com/atsign-foundation/at_server/pull/2292/commits/80b03e089322c997ff6a4187706c633d72766c54))
- published at_persistence_spec and at_persistence_secondary_server via other PRs, and removed dependency overrides from at_secondary_server ([commit](https://github.com/atsign-foundation/at_server/pull/2292/commits/32d938b7e780f28b5d27e59da671699357b25ae1))
- added a cache to EnrollmentManager ([commit](https://github.com/atsign-foundation/at_server/pull/2292/commits/4d17e9f781e098fb299e01fd65a336a3ef570e03))

Please re-review - thank you!

xavierchanth reviewed (approved):
I assume that the failed test is a flake, it doesn't look like it's related
